man/ifsysctl.5.in
.\" Automatically generated by Pandoc
.\"
.TH "IFSYSCTL" "5" "October 2024" "Wicked User Manual" ""
.SH NAME
ifsysctl[\-\f[CR]<interface name>\f[R]] \- per network interface sysctl
settings
.SH SYNOPSIS
.SS \f[B]system sysctl\f[R] files:
.IP
.EX
/boot/sysctl.conf\-<kernelversion>
/run/sysctl.d/*.conf
/etc/sysctl.d/*.conf
/usr/local/lib/sysctl.d/*.conf
/usr/lib/sysctl.d/*.conf
/lib/sysctl.d/*.conf
/etc/sysctl.conf
.EE
.SS \f[B]wicked ifsysctl\f[R] files (deprecated):
.IP
.EX
/etc/sysconfig/network/ifsysctl
/etc/sysconfig/network/ifsysctl\-\[ga]<interface>\[ga]
.EE
.SH DESCRIPTION
The settings from the \f[B]system sysctl\f[R] files are applied by
\f[CR]systemd\-sysctl.service\f[R] at boot as well as the per\-interface
settings by udev rules when the interface appears in the system.
.PP
The following configuration sets are supported by the kernel:
.IP \[bu] 2
\f[B]all\f[R]: common runtime behavior policy for all currently existing
interfaces.
.IP \[bu] 2
\f[B]default\f[R]: template used to initialize the interface sysctls,
when a new interface is created
.IP \[bu] 2
\f[B]\f[CB]<interface>\f[B]\f[R]: interface specific behavior
.PP
Wicked is never applying the \f[B]all\f[R] or \f[B]default\f[R] sysctl
settings to the kernel, but considers them and the
\f[B]\f[CB]<interface>\f[B]\f[R] sysctl settings (with highest
priority), when parsing the \f[CR]ifcfg\-<interface>\f[R] files.
.PP
The ifcfg variables and the sysctl settings are used as protocol
settings in the effective wicked xml configuration, visible in the
\f[CR]\[aq]wicked show\-config\[aq]\f[R] output.
.PP
The wicked xml interface configuration is sent by
\f[CR]\[aq]wicked ifup\[aq]\f[R] and
\f[CR]\[aq]wicked ifreload\[aq]\f[R] to the \f[CR]wickedd*\f[R]
backends, which are applying the sysctl settings from the xml
configuration to the kernel on configuration changes, hotplug events and
\f[CR]wickedd\f[R] service restart (e.g.\ on wicked package update).
.PP
Interface sysctl settings initialized by the kernel to \f[B]\-1\f[R]
(not applicable for this interface) are not overwritten by
\f[B]wickedd\f[R].
.SH SYSCTL PITFALLS
.IP \[bu] 2
Each sysctl has its own specific processing in the kernel.
For details, please read the kernel documentation
\f[CR]https://www.kernel.org/doc/Documentation/networking/ip\-sysctl.rst\f[R].
.IP \[bu] 2
Applying changes to the \f[B]all\f[R] sysctl set (e.g.\ using
\f[CR]\[aq]sysctl \-w \&...\[aq]\f[R]) is propagated (in many cases) by
the kernel in time of execution also to \f[B]default\f[R] and the
interface specific sysctl of all existing interfaces.
.IP \[bu] 2
Enabling forwarding changes the system role from host to router.
This has impact on several other sysctl\[cq]s like the IPv6 RA
processing (see accept_ra) and IPv6 autoconfiguration:
.RS 2
.IP \[bu] 2
there is no automatic IPv6 default route configuration
.IP \[bu] 2
there is no automatic IPv6 route to the prefix network
.IP \[bu] 2
there are no IPv6 addresses applied by SLAAC and DHCPv6
.IP \[bu] 2
the nameservers sent in the RA are not applied to resolv.conf
.PP
The \f[B]\f[CB]accept_ra = 2\f[B]\f[R] sysctl (on affected uplink
interface) enables RA processing on a router.
.RE
.IP \[bu] 2
Wicked is considering \f[B]system sysctl\f[R] configuration and may
revert temporary runtime changes
(\f[CR]\[aq]sysctl \-w \&...\[aq]\f[R]), e.g.\ applied by another
programs (see \f[I]DESCRIPTION\f[R]).
.SH SYSCTL VARIABLES
.SS IPv4
.TP
\f[CR]net.ipv4.conf.{all,default,interface}.forwarding\f[R]
Enable IP forwarding changing the system role between host and router.
.TP
\f[CR]net.ipv4.conf.{all,default,interface}.arp_notify\f[R]
Define mode for notification of address and device changes.
This setting is also set by \f[B]SEND_GRATUITOUS_ARP\f[R], which has
precedence over sysctl (see \f[I]man ifcfg\f[R]).
.SS IPv6:
.TP
\f[CR]net.ipv6.conf.{all,default,interface}.disable_ipv6\f[R]
Enable/Disable IPv6 operation.
.TP
\f[CR]net.ipv6.conf.{all,default,interface}.forwarding\f[R]
Enable IPv6 forwarding changing the system role between host and router.
Note: Several other sysctl\[cq]s, dhcp6, auto6 and routing have
functional dependency to the \f[CR]forwarding\f[R] sysctl (see
\f[I]SYSCTL PITFALLS\f[R]).
.TP
\f[CR]net.ipv6.conf.{all,default,interface}.accept_ra\f[R]
Accept Router Advertisements and perform IPv6 autoconfiguration of the
interface if \f[B]accept_ra > forwarding\f[R].
.TP
\f[CR]net.ipv6.conf.{all,default,interface}.autoconf\f[R]
Autoconfigure addresses using Prefix Information in Router
Advertisements.
.TP
\f[CR]net.ipv6.conf.{all,default,interface}.use_tempaddr\f[R]
Preference for Privacy Extensions (RFC3041), \f[I]ignored\f[R] for
loopback interface.
.TP
\f[CR]net.ipv6.conf.{all,default,interface}.accept_dad\f[R]
Whether to accept DAD (Duplicate Address Detection), \f[I]ignored\f[R]
for loopback interface.
.TP
\f[CR]net.ipv6.conf.{interface}.accept_redirects\f[R]
Accept Redirects.
Wicked considers only interface specific settings.
.TP
\f[CR]net.ipv6.conf.{all,default,interface}.addr_gen_mode\f[R]
Defines how link\-local and managed autoconf addresses are generated.
.TP
\f[CR]net.ipv6.conf.{all,default,interface}.stable_secret\f[R]
This IPv6 address will be used as a secret to generate IPv6 addresses
for link\-local and autoconfigured addresses.
.SH SYNTAX
The sysctl configuration supports two separator characters for sysctl
keywords: a \[lq]\f[B].\f[R]\[rq] in \f[I]default format\f[R] and a
\[lq]\f[B]/\f[R]\[rq] in the \f[I]alternate format\f[R].
Therefore, the syntax is the same as is found in the
\f[I]/etc/sysctl.conf\f[R] file.
.SH EXAMPLES
Settings for \[lq]eth0\[rq] and \[lq]eth0.1\[rq] interfaces
.IP
.EX
# using \[dq].\[dq] as separator:
net.ipv6.conf.eth0.use_tempaddr = 2
net.ipv6.conf.eth0/1.use_tempaddr = 2

# using \[dq]/\[dq] as separator:
net/ipv6/conf/eth0/use_tempaddr = 2
net/ipv6/conf/eth0.1/use_tempaddr = 2
.EE
.SH COPYRIGHT
Copyright (C) 2024 SUSE LLC
.SH BUGS
Please report bugs as described at <@PACKAGE_BUGREPORT@>
.SH SEE ALSO
\f[B]\f[CB]ifup\f[B]\f[R] (8), \f[B]\f[CB]ifcfg\f[B]\f[R] (5),
\f[B]\f[CB]sysctl\f[B]\f[R] (8), \f[B]\f[CB]systemd\-sysctl\f[B]\f[R]
(8)
.SH AUTHORS
Marius Tomaschewski, Clemens Famulla\-Conrad.


ChangeLog
#
# This file is generated while "make dist" from git repository
#


client/suse/ifsysctl.c
/*
 *	wicked client utilities to parse sysctl/ifsysctl files.
 *
 *	Copyright (C) 2011-2014 SUSE LINUX Products GmbH, Nuernberg, Germany.
 *
 *	This program is free software; you can redistribute it and/or modify
 *	it under the terms of the GNU General Public License as published by
 *	the Free Software Foundation; either version 2 of the License, or
 *	(at your option) any later version.
 *
 *	This program is distributed in the hope that it will be useful,
 *	but WITHOUT ANY WARRANTY; without even the implied warranty of
 *	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *	GNU General Public License for more details.
 *
 *	You should have received a copy of the GNU General Public License along
 *	with this program; if not, see <http://www.gnu.org/licenses/> or write
 *	to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
 *	Boston, MA 02110-1301 USA.
 *
 *	Authors:
 *		Marius Tomaschewski <mt@suse.de>
 *
 */
#if defined(HAVE_CONFIG_H)
#include <config.h>
#endif
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>
#include <errno.h>
#include <ctype.h>

#include <wicked/logging.h>
#include "ifsysctl.h"

static char *
__ni_string_strip_spaces(char *string)
{
	char *end;

	if (ni_string_empty(string))
		return string;

	/* strip trailing spaces */
	end = string + strlen(string);
	while (end-- > string) {
		if (!isspace((unsigned char)*end))
			break;
		*end = '\0';
	}

	/* strip leading spaces */
	while (isspace((unsigned char)*string))
		string++;

	return string;
}

void
__ni_sysctl_rewrite_to_slash(char *ptr)
{
	char *p, sep = 0;

	if(!ptr)
		return;

	for (p=ptr; *p; p++) {
		switch (*p) {
		case '/':
			switch (sep) {
			case 0:
				sep = '/';
				break;
			case '/':
				break;
			default:
				*p = '.';
				break;
			}
			break;
		case '.':
			switch (sep) {
			case 0:
				sep = '.';
				*p = '/';
				break;
			case '.':
				*p = '/';
				break;
			default:
				break;
			}
		default:
			break;
		}
	}
}

void
__ni_sysctl_rewrite_to_dot(char *ptr)
{
	char *p, sep = 0;

	if(!ptr)
		return;

	for (p=ptr; *p; p++) {
		switch (*p) {
		case '/':
			switch (sep) {
			case 0:
				sep = '/';
				*p = '.';
				break;
			case '/':
				*p = '.';
				break;
			default:
				break;
			}
			break;
		case '.':
			switch (sep) {
			case 0:
				sep = '.';
				break;
			case '.':
				break;
			default:
				*p = '/';
				break;
			}
		default:
			break;
		}
	}
}

ni_bool_t
__ni_sysctl_file_load(ni_var_array_t *vars, const char *filename,
		 void (*process)(ni_var_array_t *, const char *, const char *))
{
	char buffer[PATH_MAX + 1] = {'\0'};
	char *key, *val;
	FILE *fp;

	fp = fopen(filename, "re");
	if(fp == NULL) {
		if (errno != ENOENT)
			ni_error("Unable to open %s: %m", filename);
		else
			ni_debug_readwrite("Cannot open '%s': %m", filename);
		return FALSE;
	}

	while (fgets(buffer, sizeof(buffer), fp) != NULL) {
		key = __ni_string_strip_spaces(buffer);
		if (*key == '\0' || *key == ';' || *key == '#')
			continue;

		if ((val = strchr(key, '=')) == NULL)
			continue;
		*val++ = '\0';

		key = __ni_string_strip_spaces(key);
		val = __ni_string_strip_spaces(val);

		if(*key && *val) {
			__ni_sysctl_rewrite_to_dot(key);
			process(vars, key, val);
		}
	}

	fclose(fp);
	return TRUE;
}

ni_var_t *
ni_ifsysctl_vars_get(const ni_var_array_t *vars, const char *keyfmt, ...)
{
	ni_var_t *var;
	char *key = NULL;
	va_list ap;
	int ret;

	if(!vars || !keyfmt)
		return NULL;

	va_start(ap, keyfmt);
	ret = vasprintf(&key, keyfmt, ap);
	va_end(ap);
	if (ret < 0)
		return NULL;

	__ni_sysctl_rewrite_to_dot(key);
	var = ni_var_array_get(vars, key);
	ni_string_free(&key);
	return var;
}

ni_bool_t
ni_ifsysctl_vars_set(ni_var_array_t *vars, const char *value, const char *keyfmt, ...)
{
	char *key = NULL;
	va_list ap;
	int ret;

	if(!vars || !keyfmt || !value)
		return FALSE;

	va_start(ap, keyfmt);
	ret = vasprintf(&key, keyfmt, ap);
	va_end(ap);
	if (ret < 0)
		return FALSE;

	__ni_sysctl_rewrite_to_dot(key);
	ni_var_array_set(vars, key, value);
	ni_string_free(&key);
	return TRUE;
}

void
__ni_ifsysctl_vars_map(ni_var_array_t *vars, const char *key, const char *val)
{
	ni_stringbuf_t buf = NI_STRINGBUF_INIT_DYNAMIC;
	const char *ptr;

	/* Normalize the net.ipv4.ip_forward alias */
	if (!strcmp(key, "net.ipv4.ip_forward"))
		key = "net.ipv4.conf.all.forwarding";

	/*
	 * Filter out net.ipv4.conf.* and net.ipv6.conf.* only.
	 */
	if (strncmp(key, "net.ipv4.conf.", sizeof("net.ipv4.conf.")-1)
	&&  strncmp(key, "net.ipv6.conf.", sizeof("net.ipv6.conf.")-1))
		return;

	/*
	 * Resolve $INTERFACE and $SYSCTL_IF wildcard crap
	 */
	if ((ptr = strstr(key, "$INTERFACE"))) {
		ni_stringbuf_puts(&buf, key);
		ni_stringbuf_truncate(&buf, ptr - key);
		ni_stringbuf_puts(&buf, "default");
		ptr += sizeof("$INTERFACE")-1;
		ni_stringbuf_puts(&buf, ptr);
		key = buf.string;
	} else
	if ((ptr = strstr(key, "$SYSCTL_IF"))) {
		ni_stringbuf_puts(&buf, key);
		ni_stringbuf_truncate(&buf, ptr - key);
		ni_stringbuf_puts(&buf, "default");
		ptr += sizeof("$SYSCTL_IF")-1;
		ni_stringbuf_puts(&buf, ptr);
		key = buf.string;
	}

	/*
	 * And finally add it to the array
	 */
	ni_debug_readwrite("Add sysctl variable '%s=%s'", key, val);
	ni_var_array_set(vars, key, val);
	ni_stringbuf_destroy(&buf);
}

ni_bool_t
ni_ifsysctl_file_load(ni_var_array_t *vars, const char *filename)
{
	if (!vars || ni_string_empty(filename))
		return FALSE;

	ni_debug_readwrite("Reading sysctl file '%s'", filename);
	return __ni_sysctl_file_load(vars, filename, __ni_ifsysctl_vars_map);
}
